evaluation/EDA/game.py

```diff
@@ -0,0 +1,370 @@
+            roles = ['assistant', 'user']
+            for i, message in enumerate(raw_messages):
+                self.guesser_messages.append(
+                    {'role': roles[i % 2], 'content': message['content']}
+                )
+
+        self.guesser_messages = self.guesser_messages[:-2]
+        self.guesser_messages[-1]['content'] = (
+            self.guesser_messages[-1]['content'] + " You must guess now, what's it?"
+        )
+        guesser_msg = self.guesser(self.guesser_messages)
+        self.guesser_messages.append(guesser_msg)
+        guesser_question = guesser_msg['content'].strip()
+        self.guesser_messages[-1]['content'] = (
+            self.guesser_messages[-1]['content'] + ' Is it right?'
+        )
+        usr_msg = self.answerer(guesser_question)
+        self.guesser_messages.append(
+            {'role': 'user', 'content': f"{usr_msg['content'].strip()}"}
+        )
+
+        if 'bingo' in self.guesser_messages[-1]['content'].lower():
+            self.guesser_win = True
+            return True
+
+        return False
+
+    @retry(
+        (
+            openai.Timeout,
+            requests.exceptions.ReadTimeout,
+            openai.RateLimitError,
+            openai.APIError,
+            requests.exceptions.HTTPError,
+            openai.APIConnectionError,
+        ),
+        tries=5,
+        delay=0.5,
+        backoff=0.5,
+        max_delay=2,
+        logger=LOGGER,
+    )
+    def guesser(self, messages):
+        if not self.guesser_model.startswith('gpt'):  # hf model
+            self.guesser_model, self.guesser_tokenizer = load_model(self.guesser_model)
+
+            # """Wraps hf's `generate` adding some specific method's defaults"""
+            assert not self.openai_api
+            prompt = self.dialog_history() + ' ASSISTANT:'
+            input_ids = torch.tensor(
+                [self.guesser_tokenizer.encode(prompt, add_special_tokens=True)]
+            )  # TODO check if huggingface is using the same format.
+            input_ids = input_ids.to(self.guesser_model.base_model.device)
+            attention_mask = None
+
+            with torch.no_grad():
+                gen = self.guesser_model.generate(
+                    input_ids=input_ids,
+                    attention_mask=attention_mask,
+                    **self.guesser_kargs,
+                )
+                gen_str = (
+                    self.guesser_tokenizer.decode(gen[0][input_ids[0].shape[0] :])
+                    .split('</s>')[0]
+                    .split('USER')[0]
+                    .lstrip()
+                    .strip()
+                )
+
+                return {
+                    'role': 'assistant',
+                    'content': gen_str,
+                }
+        else:
+            openai.api_base = self.guesser_api_base
+            client = OpenAI(api_key=openai.api_key)
+            response = client.chat.completions.create(
```

Comment:
Maybe we can replace all the OpenAI API calls with litellm, and log costs properly as introduced here: https://github.com/OpenDevin/OpenDevin/pull/1944

@yufansong could you help take a look?

Reply:
Once this is done, and the eval can run end-to-end (i can help verify)- i think this is good to merge

Reply:
Thanks @xingyaoww @yufansong